### DISCUSSION TITLE
任务队列问题

### DISCUSSION THREAD

**Discussion by JieTrancender** [cloudwu/skynet#1345] [Q&A]
问题起源：某些对全服玩家遍历然后执行的操作会导致服务被单个协程长时间占用并触发`maybe endless loop`警告。
根据前面云大的评论参照https://github.com/cloudwu/skynet/blob/master/lualib/snax/gateserver.lua#L65 改了一版

~~~ lua
-- 消费任务
function TaskQueueMgr:consumeTask()
    if not next(self._taskQueue) then
        self._count = self._count - 1
        return
    end

    -- 处理消息直到消息阻塞
    -- 如果没有消息阻塞，队列就会为空，所以fork一次然后结束就可以了
    skynet.fork(self.consumeTask, self)
    self._count = self._count + 1

    local f, ok, task
    task = table_remove(self._taskQueue, 1)
    f = task[2]
    if type(f) == "function" then
        ok = xpcall(f, traceback, task[1], table_unpack(task[3]))
        if not ok then
            logError("TaskQueueMgr#consumeTask fail", task[1], task[2], util.table_dump_line(task[3]))
        end
    else
        logError("TaskQueueMgr#consumeTask fail, is not function", task[1], task[2], util.table_dump_line(task[3]))
    end

    self._count = self._count - 1
end
~~~
此时报警不会再出现，但是外部新请求依然会排队很久才能执行(因为去执行全服玩家遍历操作的任务协程排在前面)。

为了解决这个问题采用了折中的方法：**新起一个任务队列服务，当前服务将遍历玩家需要执行的操作发送到任务队列服务中去然后自己再一个一个任务处理，此时外部请求和处理任务队列里面的任务优先级一致**。
代码见：https://gist.github.com/JieTrancender/2b3deb76671a5a38f7124d98d6d50f2e

但是当前解决方法需要服务发送一次请求去队列服务并且新增接口供任务被消耗，请问有更好的方法实现这个需求吗？

**Comment by cloudwu** [comment]
我觉得你要的是尽量不干扰外部消息。所以，在你处理自己的内部任务的间隙，判断一下 mqlen 的长度，然后如果有外部消息调用 skynet.yield 就够了。

**Reply by JieTrancender** [reply]
是的，主要述求就是不干扰外部消息。这个方案可行，谢谢。

**Reply by JieTrancender** [reply]
但是还是希望有类似消息队列的处理办法，当类似的操作变多的时候，一直切出去会导致很多协程产生。
